File: demo-service-impl/src/test/java/com/github/zubmike/service/demo/logic/StarshipLogicTest.java
package com.github.zubmike.service.demo.logic;

import com.github.zubmike.core.utils.DateTimeUtils;
import com.github.zubmike.core.utils.DuplicateException;
import com.github.zubmike.core.utils.InvalidParameterException;
import com.github.zubmike.core.utils.NotFoundException;
import com.github.zubmike.service.demo.api.types.StarshipEntry;
import com.github.zubmike.service.demo.api.types.StarshipInfo;
import com.github.zubmike.service.demo.dao.PlanetarySystemDao;
import com.github.zubmike.service.demo.dao.StarshipDao;
import com.github.zubmike.service.demo.types.PlanetarySystem;
import com.github.zubmike.service.demo.types.ServiceUserContext;
import com.github.zubmike.service.demo.types.Starship;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.*;

import java.time.LocalDateTime;
import java.util.Locale;
import java.util.Optional;

public class StarshipLogicTest {

	private static final PlanetarySystem TEST_PLANET_SYSTEM = new PlanetarySystem(1, "Test system", "TST");

	@Spy
	private ServiceUserContext serviceUserContext = new ServiceUserContext(1, Locale.getDefault());
	@Mock
	private StarshipDao starshipDao;
	@Mock
	private PlanetarySystemDao planetarySystemDao;

	@InjectMocks
	private StarshipLogic starshipLogic;

	@Before
	public void init() {
		MockitoAnnotations.initMocks(this);
		Mockito.when(planetarySystemDao.get(Mockito.eq(TEST_PLANET_SYSTEM.getId())))
				.thenReturn(Optional.of(TEST_PLANET_SYSTEM));
		Mockito.when(planetarySystemDao.getByCode(Mockito.eq(TEST_PLANET_SYSTEM.getCode())))
				.thenReturn(Optional.of(TEST_PLANET_SYSTEM));
	}

	@Test
	public void addStarship() {
		starshipLogic.addStarship(new StarshipEntry("TST-000001"));
	}

	@Test(expected = DuplicateException.class)
	public void addDuplicateStarship() {
		Mockito.when(starshipDao.getByNumber(Mockito.eq("TST-000002")))
				.thenAnswer(invocation -> {
					var item = new Starship();
					item.setId(0L);
					item.setNumber(invocation.getArgument(0));
					return Optional.of(item);
				});
		starshipLogic.addStarship(new StarshipEntry("TST-000002"));
	}

	@Test(expected = InvalidParameterException.class)
	public void addStarshipWithInvalidNumber() {
		starshipLogic.addStarship(new StarshipEntry("123"));
	}

	@Test(expected = InvalidParameterException.class)
	public void addStarshipFromUnknownPlanet() {
		starshipLogic.addStarship(new StarshipEntry("QWE-000001"));
	}

	@Test
	public void getStarship() {
		Starship mockStarship = new Starship();
		mockStarship.setId(3L);
		mockStarship.setNumber("TST-000003");
		mockStarship.setPlanetarySystemId(TEST_PLANET_SYSTEM.getId());
		mockStarship.setTimeCount(99);
		mockStarship.setCreateDate(LocalDateTime.now());

		Mockito.when(starshipDao.get(Mockito.eq(3L)))
				.thenReturn(Optional.of(mockStarship));
		Mockito.when(planetarySystemDao.getName(Mockito.eq(TEST_PLANET_SYSTEM.getId())))
				.thenReturn(TEST_PLANET_SYSTEM.getName());

		StarshipInfo starshipInfo = starshipLogic.getStarship(3L);

		Assert.assertEquals(mockStarship.getId().longValue(), starshipInfo.getId());
		Assert.assertEquals(mockStarship.getNumber(), starshipInfo.getNumber());
		Assert.assertEquals(mockStarship.getPlanetarySystemId(), starshipInfo.getPlanetarySystemId());
		Assert.assertEquals(TEST_PLANET_SYSTEM.getName(), starshipInfo.getPlanetarySystemName());
		Assert.assertEquals(mockStarship.getTimeCount(), starshipInfo.getTimeCount());
		Assert.assertEquals(DateTimeUtils.toString(mockStarship.getCreateDate()), starshipInfo.getCreateDate());
	}

	@Test(expected = NotFoundException.class)
	public void getUnknownStarship() {
		starshipLogic.getStarship("TST-999999");
	}
}

File: demo-service-kotlin-impl/src/test/kotlin/ru/zubmike/service/demo/logic/StarshipLogicTest.kt
package ru.zubmike.service.demo.logic

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import ru.zubmike.core.utils.DateTimeUtils
import ru.zubmike.core.utils.InvalidParameterException
import ru.zubmike.core.utils.NotFoundException
import ru.zubmike.service.demo.api.types.StarshipEntry
import ru.zubmike.service.demo.dao.PlanetarySystemDao
import ru.zubmike.service.demo.dao.StarshipDao
import ru.zubmike.service.demo.types.PlanetarySystem
import ru.zubmike.service.demo.types.Starship
import ru.zubmike.service.demo.utis.MockitoExt
import java.time.LocalDateTime
import java.util.*

class StarshipLogicTest {

    companion object {

        private val TEST_PLANET_SYSTEM = PlanetarySystem(1, "Test system", "TST")

    }

    @Mock
    private lateinit var starshipDao: StarshipDao

    @Mock
    private lateinit var planetarySystemDao: PlanetarySystemDao

    @InjectMocks
    private lateinit var starshipLogic: StarshipLogic

    @Before
    fun init() {
        MockitoAnnotations.initMocks(this)
        MockitoExt.whenever(planetarySystemDao.get(MockitoExt.safeEq<Int>(TEST_PLANET_SYSTEM.id)))
                .thenReturn(Optional.of(TEST_PLANET_SYSTEM))
        MockitoExt.whenever(planetarySystemDao.getByCode(MockitoExt.safeEq(TEST_PLANET_SYSTEM.code)))
                .thenReturn(TEST_PLANET_SYSTEM)
    }

    @Test
    fun addStarship() {
        starshipLogic.addStarship(StarshipEntry("TST-000001"))
    }

    @Test(expected = InvalidParameterException::class)
    fun addDuplicateStarship() {
        MockitoExt.whenever(starshipDao.getByNumber(MockitoExt.safeEq("TST-000002")))
                .thenAnswer { invocation ->
                    val item = Starship()
                    item.id = 0
                    item.number = invocation.getArgument<String>(0)
                    item
                }
        starshipLogic.addStarship(StarshipEntry("TST-000002"))
    }

    @Test(expected = InvalidParameterException::class)
    fun addStarshipWithInvalidNumber() {
        starshipLogic.addStarship(StarshipEntry("123"))
    }

    @Test(expected = InvalidParameterException::class)
    fun addStarshipFromUnknownPlanet() {
        starshipLogic.addStarship(StarshipEntry("QWE-000001"))
    }

    @Test
    fun getStarship() {
        val mockStarship = Starship()
        mockStarship.id = 3L
        mockStarship.number = "TST-000003"
        mockStarship.planetarySystemId = TEST_PLANET_SYSTEM.id
        mockStarship.timeCount = 99
        mockStarship.createDate = LocalDateTime.now()

        MockitoExt.whenever(starshipDao.get(MockitoExt.safeEq(3L)))
                .thenReturn(Optional.of(mockStarship))
        MockitoExt.whenever(planetarySystemDao.getName(MockitoExt.safeEq<Int>(TEST_PLANET_SYSTEM.id)))
                .thenReturn(TEST_PLANET_SYSTEM.name)

        val starshipInfo = starshipLogic.getStarship(3L)

        Assert.assertEquals(mockStarship.id, starshipInfo.id)
        Assert.assertEquals(mockStarship.number, starshipInfo.number)
        Assert.assertEquals(mockStarship.planetarySystemId.toLong(), starshipInfo.planetarySystemId.toLong())
        Assert.assertEquals(TEST_PLANET_SYSTEM.name, starshipInfo.planetarySystemName)
        Assert.assertEquals(mockStarship.timeCount.toLong(), starshipInfo.timeCount.toLong())
        Assert.assertEquals(DateTimeUtils.toString(mockStarship.createDate), starshipInfo.createDate)
    }

    @Test(expected = NotFoundException::class)
    fun getUnknownStarship() {
        starshipLogic.getStarship("TST-999999")
    }

}
